plugin-server/src/cdp/async-function-executor.ts

```diff
@@ -0,0 +1,103 @@
+
+export class AsyncFunctionExecutor {
+    constructor(
+        private serverConfig: PluginsServerConfig,
+        private rustyHook: RustyHook,
+        private kafkaProducer: KafkaProducerWrapper
+    ) {}
+
+    async execute(request: HogFunctionInvocationAsyncRequest): Promise<HogFunctionInvocationAsyncRequest> {
+        const loggingContext = {
+            hogFunctionId: request.hogFunctionId,
+            invocationId: request.id,
+            asyncFunctionName: request.asyncFunctionName,
+        }
+        status.info('🦔', `[AsyncFunctionExecutor] Executing async function`, loggingContext)
+
+        switch (request.asyncFunctionName) {
+            case 'fetch':
+                await this.asyncFunctionFetch(request)
+                break
+            default:
+                status.error('🦔', `[HogExecutor] Unknown async function: ${request.asyncFunctionName}`, loggingContext)
+        }
+
+        return request
+    }
+
+    private async asyncFunctionFetch(request: HogFunctionInvocationAsyncRequest): Promise<any> {
+        // TODO: validate the args
+        const args = request.asyncFunctionArgs ?? []
+        const url: string = args[0]
+        const options = args[1]
+
+        const method = options.method || 'POST'
+        const headers = options.headers || {
+            'Content-Type': 'application/json',
+        }
+        const body = options.body || {}
+
+        const webhook: Webhook = {
+            url,
+            method: method,
+            headers: headers,
+            body: typeof body === 'string' ? body : JSON.stringify(body, undefined, 4),
+        }
+
+        // NOTE: Purposefully disabled for now - once we have callback support we can re-enable
+        // const SPECIAL_CONFIG_ID = -3 // Hardcoded to mean Hog
+        // const success = await this.rustyHook.enqueueIfEnabledForTeam({
+        //     webhook: webhook,
+        //     teamId: hogFunction.team_id,
+        //     pluginId: SPECIAL_CONFIG_ID,
+        //     pluginConfigId: SPECIAL_CONFIG_ID,
+        // })
+
+        const success = false
+
+        // TODO: Temporary test code
+        if (!success) {
+            status.info('🦔', `[HogExecutor] Webhook not sent via rustyhook, sending directly instead`)
+            const response: HogFunctionInvocationAsyncResponse = {
+                ...request,
+            }
+
+            try {
+                const fetchResponse = await trackedFetch(url, {
+                    method: webhook.method,
+                    body: webhook.body,
+                    headers: webhook.headers,
+                    timeout: this.serverConfig.EXTERNAL_REQUEST_TIMEOUT_MS,
+                })
+
+                const maybeJson = await fetchResponse.json().catch(() => null)
+
+                response.vmResponse = {
+                    status: fetchResponse.status,
+                    body: await fetchResponse.text(),
+                    json: maybeJson,
```

Comment:
If we're sending this through Kafka, having both `body` and `json` will likely double the message size... so perhaps `maybeJson ? null : (await fetchResponse.text())` or something like that?

Reply:
Yeah that makes sense